reactor-core/src/main/java/reactor/core/publisher/FluxOnErrorReturn.java

```diff
@@ -0,0 +1,231 @@
+				s.cancel();
+			}
+		}
+
+		@Override
+		public void request(long n) {
+			long r;
+			for (;;) {
+				r = REQUESTED.get(this);
+				if (r == Long.MAX_VALUE || r < STATE_PENDING_FALLBACK) {
+					return;
+				}
+
+				if (r == STATE_PENDING_FALLBACK) {
+					//ensure no concurrent request took care of it already
+					if (REQUESTED.compareAndSet(this, r, STATE_TERMINATED)) {
+						if (this.fallbackValue != null) { //should be always true at this point
+							actual.onNext(this.fallbackValue);
+						}
+						actual.onComplete();
+						//optionally still transmit the request to upstream + cancel it
+						// to indicate eg. stress tests that we went down this code path
+						if (this.trackRequestWhenFallbackDeferred) {
+							s.request(n);
+							s.cancel();
+						}
+					}
+					return;
+				}
+
+				long u = Operators.addCap(r, n);
+				if (REQUESTED.compareAndSet(this, r, u)) {
+					s.request(n);
+					return;
+				}
+			}
+		}
+
+		@Override
+		public void onNext(T t) {
+			producedOne();
+			actual.onNext(t);
+		}
+
+		@Override
+		public void onComplete() {
+			if (this.requested != STATE_CANCELLED) {
+				REQUESTED.set(this, STATE_TERMINATED);
+				actual.onComplete();
+			}
+		}
+
+		@Override
+		public void onError(Throwable t) {
+			boolean shouldResume = this.resumableErrorPredicate == null || this.resumableErrorPredicate.test(t);
+			if (!shouldResume) {
+				//it is always fine to propagate errors even when there is no pending request
+				if (this.requested != STATE_CANCELLED) {
+					REQUESTED.set(this, STATE_TERMINATED);
+					actual.onError(t);
+				}
+				return;
+			}
+
+			if (this.fallbackValue == null) {
+				//it is always fine to send onComplete() even when there is no pending request
+				if (this.requested != STATE_CANCELLED) {
+					REQUESTED.set(this, STATE_TERMINATED);
+					actual.onComplete();
+				}
+				return;
+			}
+
+//			in case of a fallback value, we need to make sure we don't send onNext(fallback) if there is no request
+			long r = this.requested;
+			if (r > 0) {
+				//upstream won't produce anymore so we're confident the fallback value won't compete with an onNext for demand
+				REQUESTED.set(this, STATE_TERMINATED);
+				actual.onNext(this.fallbackValue);
+				actual.onComplete();
+				return;
+			}
+
+			//now the interesting case: r == 0 means we MUST wait for the next request before delivering the fallback value
+			//upstream still won't produce, so the requested won't further diminish via producedOne()
+			//we try to turn that 0 into a marker for pending fallback but if we fail we might want to propagate fallback ourselves
+			if (!REQUESTED.compareAndSet(this, 0, STATE_PENDING_FALLBACK)) {
+				//re-check the request:
+				r = this.requested;
+				if (r > 0) {
+					REQUESTED.set(this, STATE_TERMINATED);
+					actual.onNext(this.fallbackValue);
+					actual.onComplete();
+				}
+			} // else next request will take care of delivering
```

Comment:
If `r <= 0` in the inner condition above, the only option is that it was cancelled, correct? Perhaps worth adding a comment about it?

Reply:
good idea !